### PR TITLE
Combine `signif` into round, use keyword args for digits/sigdigits

### PR DESCRIPTION
**Pull request by simonbyrne**
This is my attempt at fixing #26663.

Internally, this ended up being a bit cleaner, as they share a common codebase. The only downside is that dispatch is slightly more complicated (e.g. to define `round` for a new numeric type, you need to define a method for `Base._round`, which is a bit cumbersome).

Looking at this a bit more, we could actually generalise this a bit further, by adding 2 more:

* `step` or `increment` keyword, which would be equivalent to `round(x, digits=-1, base=step)`. This would correspond to [how `round` currently behaves with dates](https://docs.julialang.org/en/stable/manual/dates/#Rounding-1) (for dates we could also add an extra `epoch` keyword).
* an `nth` or `inversestep` keyword (there is probably a better name), which would be equivalent to `round(x, digits=1, base=nth)`.

In this way, `round(x, digits=..)` could in turn call these.

### CODE REVIEW

File: base/floatfuncs.jl [hunk 1]
```diff
@@ -44,43 +44,51 @@ isinteger(x::AbstractFloat) = (x - trunc(x) == 0)
 
 """
     round([T,] x, [r::RoundingMode])
-    round(x, [digits; base = 10])
+    round(x, [r::RoundingMode]; digits::Integer= [, base = 10])
```
Comment (nalimilan) [inline base/floatfuncs.jl]: `digits::Integer=0, base::Integer=10`?

Comment (simonbyrne) [reply base/floatfuncs.jl]: I thought about that, but the default `digits = 0` is covered by the above.

Comment (nalimilan) [reply base/floatfuncs.jl]: Then just omit `=`. Though I'm not sure it's a problem to repeat the default. Also brackets are never used for keyword arguments.

BTW, below there's a typo in "provied".

**Comment by simonbyrne** [comment]
@KristofferC Have you seen this error before?
https://freebsdci.julialang.org/#/builders/1/builds/8653/steps/7/logs/stdio

**Comment by KristofferC** [comment]
@simonbyrne https://github.com/JuliaLang/julia/pull/26705

File: base/float.jl [hunk 2]
```diff
@@ -662,7 +660,7 @@ for Ti in (Int8, Int16, Int32, Int64, Int128, UInt8, UInt16, UInt32, UInt64, UIn
                     end
                 end
                 function (::Type{$Ti})(x::$Tf)
-                    if ($(Tf(typemin(Ti))) <= x <= $(Tf(typemax(Ti)))) && (trunc(x) == x)
+                    if ($(Tf(typemin(Ti))) <= x <= $(Tf(typemax(Ti)))) && (_round(x, RoundToZero) == x)
```
Comment (JeffBezanson) [inline base/float.jl]: Curious: why this change?

**Comment by JeffBezanson** [comment]
Triage accepts.

File: base/float.jl (hunk 2 above)
Comment (simonbyrne) [reply base/float.jl]: problems with bootstrapping: `trunc(x)` had to be defined later.

**Comment by martinholters** [comment]
I think this broke
```julia
julia> round(pi, sigdigits=3, base=2)
ERROR: MethodError: no method matching exponent(::Irrational{:π})
Closest candidates are:
  exponent(::Missing) at missing.jl:75
  exponent(::BigFloat) at mpfr.jl:803
  exponent(::T<:Union{Float16, Float32, Float64}) where T<:Union{Float16, Float32, Float64} at math.jl:590
Stacktrace:
 [1] hidigit(::Irrational{:π}, ::Int64) at ./floatfuncs.jl:179
 [2] _round at ./floatfuncs.jl:186 [inlined]
 [3] #round#557 at ./floatfuncs.jl:127 [inlined]
 [4] #round at ./<missing>:0 [inlined] (repeats 2 times)
 [5] top-level scope
```
(Other bases work.)
